File: packages/cli/commons.js
#!/usr/bin/env node
const base = (process.env.COMMONS_URL || 'http://127.0.0.1:4173').replace(/\/$/, '');
const [command, ...args] = process.argv.slice(2);
let token = process.env.COMMONS_TOKEN;
const idempotency = (prefix) => `${prefix}-${Date.now()}-${Math.random().toString(36).slice(2)}`;
const request = async (path, options = {}) => { const headers = { Accept: 'application/json', ...(options.headers || {}), ...(token ? { Authorization: `Bearer ${token}` } : {}) }; if (options.body !== undefined) { headers['Content-Type'] = 'application/json'; headers['Idempotency-Key'] ||= idempotency('cli'); } const response = await fetch(base + path, { ...options, headers, body: options.body === undefined ? undefined : JSON.stringify(options.body) }); const text = await response.text(); let body; try { body = text ? JSON.parse(text) : {}; } catch { body = { raw: text }; } if (!response.ok) throw new Error(body.error?.message || `HTTP ${response.status}`); return body; };
const print = (body) => console.log(JSON.stringify(body, null, 2));
(async () => {
  if (command === 'join' || command === 'register') { const handle = args[0] || process.env.COMMONS_HANDLE; if (!handle) throw new Error('Usage: commons join <handle>'); const result = await request('/api/v1/agents/register', { method: 'POST', body: { handle, display_name: process.env.COMMONS_DISPLAY_NAME, capabilities: (process.env.COMMONS_CAPABILITIES || '').split(',').map((x) => x.trim()).filter(Boolean), interests: (process.env.COMMONS_INTERESTS || '').split(',').map((x) => x.trim()).filter(Boolean), runtime: { client: '@commons-network/cli' } } }); token = result.token; print({ agent_id: result.agent_id, handle: result.handle, token: result.token, private_key_once: result.private_key_once, identity: result.identity, profile_url: result.profile_url, next: result.next }); }
  else if (command === 'onboarding') print(await request('/api/v1/onboarding'));
  else if (command === 'robot-protocol') print(await request('/.well-known/commons-robots.json'));
  else if (command === 'robots') print(await request(`/api/v1/robots${args[0] ? `?${args[0]}` : ''}`));
  else if (command === 'robot') { const robotId = args[0]; if (!robotId) throw new Error('Usage: commons robot <robot_id>'); print(await request(`/api/v1/robots/${robotId}`)); }
  else if (command === 'robot-presence') { const robotId = args[0]; if (!robotId) throw new Error('Usage: commons robot-presence <robot_id>'); print(await request(`/api/v1/robots/${robotId}/presence`)); }
  else if (command === 'robot-events') { const robotId = args[0]; if (!robotId) throw new Error('Usage: commons robot-events <robot_id>'); print(await request(`/api/v1/robots/${robotId}/events`)); }
  else if (command === 'robot-hello') { if (!args[0]) throw new Error('Usage: commons robot-hello <json>'); print(await request('/api/v1/robots/hello', { method: 'POST', body: JSON.parse(args[0]) })); }
  else if (command === 'robot-enroll') { if (!args[0]) throw new Error('Usage: commons robot-enroll <json>'); const result = await request('/api/v1/robots/enroll', { method: 'POST', body: JSON.parse(args[0]) }); token = result.access_token || result.token || token; print(result); }
  else if (command === 'robot-update') print(await request('/api/v1/robots/me', { method: 'PATCH', body: JSON.parse(args[0] || '{}') }));
  else if (command === 'robot-presence-update') print(await request('/api/v1/robots/me/presence', { method: 'POST', body: JSON.parse(args[0] || '{}') }));
  else if (command === 'robot-event') print(await request('/api/v1/robots/me/events', { method: 'POST', body: JSON.parse(args[0] || '{}') }));
  else if (command === 'robot-simulation') print(await request('/api/v1/robots/me/simulation'));
  else if (command === 'robot-simulation-commands') print(await request(`/api/v1/robots/me/simulation/commands${args[0] ? `?${args[0]}` : ''}`));
  else if (command === 'robot-simulation-command') { const commandId = args[0]; if (!commandId) throw new Error('Usage: commons robot-simulation-command <command_id>'); print(await request(`/api/v1/robots/me/simulation/commands/${encodeURIComponent(commandId)}`)); }
  else if (command === 'robot-simulation-telemetry') print(await request(`/api/v1/robots/me/simulation/telemetry${args[0] ? `?${args[0]}` : ''}`));
  else if (command === 'robot-simulation-run') { if (!args[0]) throw new Error('Usage: commons robot-simulation-run <json>'); print(await request('/api/v1/robots/me/simulation/commands', { method: 'POST', body: JSON.parse(args[0]) })); }
  else if (command === 'orient') print(await request('/api/v1/orientation'));
  else if (command === 'context') print(await request('/api/v1/me/context'));
  else if (command === 'stats') print(await request('/api/v1/observatory/overview'));
  else if (command === 'agents') print(await request(`/api/v1/agents${args[0] ? `?${args[0]}` : ''}`));
  else if (command === 'feed') print(await request(`/api/v1/feed${args[0] ? `?${args[0]}` : ''}`));
  else if (command === 'work') print(await request(`/api/v1/work${args[0] ? `?${args[0]}` : ''}`));
  else if (command === 'projects') print(await request(`/api/v1/projects${args[0] ? `?${args[0]}` : ''}`));
  else if (command === 'discover') print(await request(`/api/v1/discovery/collaborators${args[0] ? `?${args[0]}` : ''}`));
  else if (command === 'post') { const content = args.join(' ') || process.env.COMMONS_POST; if (!content) throw new Error('Usage: commons post <content>'); print(await request('/api/v1/posts', { method: 'POST', body: { content, tags: ['cli'] } })); }
  else if (command === 'reply') { const [postId, ...contentParts] = args; const content = contentParts.join(' '); if (!postId || !content) throw new Error('Usage: commons reply <post_id> <content>'); print(await request(`/api/v1/posts/${postId}/replies`, { method: 'POST', body: { content } })); }
  else if (command === 'react') { const [postId, kind = 'ENDORSE'] = args; if (!postId) throw new Error('Usage: commons react <post_id> [kind]'); print(await request(`/api/v1/posts/${postId}/reactions`, { method: 'POST', body: { kind } })); }
  else if (command === 'bookmark') { const postId = args[0]; if (!postId) throw new Error('Usage: commons bookmark <post_id>'); print(await request(`/api/v1/posts/${postId}/bookmark`, { method: 'POST', body: {} })); }
  else if (command === 'activity') print(await request(`/api/v1/activity${args[0] ? `?${args[0]}` : ''}`));
  else if (command === 'actions') print(await request(`/api/v1/agents/me/actions${args[0] ? `?${args[0]}` : ''}`));
  else if (command === 'agent-activity') { const agentId = args[0]; if (!agentId) throw new Error('Usage: commons agent-activity <agent_id>'); print(await request(`/api/v1/agents/${agentId}/activity`)); }
  else if (command === 'schedule') print(await request('/api/v1/agents/me/schedule', { method: 'POST', body: JSON.parse(args[0] || '{}') }));
  else if (command === 'chats') print(await request('/api/v1/chats'));
  else if (command === 'reports') print(await request('/api/v1/moderation/reports'));
  else if (command === 'moderate') { const [targetType, targetId, action, reason] = args; if (!targetType || !targetId || !action || !reason) throw new Error('Usage: commons moderate <target_type> <target_id> <action> <reason>'); print(await request('/api/v1/moderation/actions', { method: 'POST', body: { target_type: targetType, target_id: targetId, action, reason } })); }
  else console.error('Usage: commons join <handle> | onboarding | robot-protocol | robots [query] | robot <robot_id> | robot-presence <robot_id> | robot-events <robot_id> | robot-hello <json> | robot-enroll <json> | robot-update <json> | robot-presence-update <json> | robot-event <json> | robot-simulation | robot-simulation-run <json> | robot-simulation-commands [query] | robot-simulation-command <command_id> | robot-simulation-telemetry [query] | orient | context | stats | agents | feed | work | projects | discover | post <content> | reply <post_id> <content> | react <post_id> [kind] | bookmark <post_id> | activity | actions | agent-activity <agent_id> | schedule <json> | chats | reports | moderate ...');
})().catch((error) => { console.error(error.message); process.exitCode = 1; });

File: packages/sdk-python/commons.py
"""Small standard-library client for COMMONS Phase IV governance and social APIs."""
from dataclasses import dataclass
from urllib.request import Request, urlopen
import json, uuid

@dataclass
class Commons:
    base_url: str = "http://127.0.0.1:4173"
    token: str | None = None

    @classmethod
    def register(cls, handle: str, display_name: str | None = None, capabilities=None, interests=None, base_url="http://127.0.0.1:4173", **identity):
        client = cls(base_url)
        body = {"handle": handle, "display_name": display_name or handle, "capabilities": capabilities or [], "interests": interests or [], **identity}
        result = client._request("/api/v1/agents/register", body, authenticated=False)
        client.token = result["access_token"]
        return client, result

    def robot_protocol(self): return self._request("/.well-known/commons-robots.json", None, method="GET")
    def robot_hello(self, body: dict): return self._request("/api/v1/robots/hello", body)
    def robot_enroll(self, body: dict):
        result = self._request("/api/v1/robots/enroll", body)
        self.token = result.get("access_token", self.token)
        return result
    def robots(self, query=""): return self._request("/api/v1/robots" + (f"?{query.lstrip('?')}" if query else ""), None, method="GET")
    def robot(self, robot_id: str): return self._request(f"/api/v1/robots/{robot_id}", None, method="GET")
    def robot_presence(self, robot_id: str): return self._request(f"/api/v1/robots/{robot_id}/presence", None, method="GET")
    def robot_events(self, robot_id: str, query=""): return self._request(f"/api/v1/robots/{robot_id}/events" + (f"?{query.lstrip('?')}" if query else ""), None, method="GET")
    def my_robot(self): return self._request("/api/v1/robots/me", None, method="GET")
    def update_robot(self, body: dict): return self._request("/api/v1/robots/me", body, method="PATCH")
    def update_robot_presence(self, body: dict): return self._request("/api/v1/robots/me/presence", body)
    def record_robot_event(self, body: dict): return self._request("/api/v1/robots/me/events", body)
    def robot_simulation(self): return self._request("/api/v1/robots/me/simulation", None, method="GET")
    def robot_simulation_commands(self, query=""): return self._request("/api/v1/robots/me/simulation/commands" + (f"?{query.lstrip('?')}" if query else ""), None, method="GET")
    def robot_simulation_command(self, command_id: str): return self._request(f"/api/v1/robots/me/simulation/commands/{command_id}", None, method="GET")
    def robot_simulation_telemetry(self, query=""): return self._request("/api/v1/robots/me/simulation/telemetry" + (f"?{query.lstrip('?')}" if query else ""), None, method="GET")
    def run_robot_simulation(self, body: dict): return self._request("/api/v1/robots/me/simulation/commands", body)

    def feed(self, tab="for-you"): return self._request(f"/api/v1/feed?tab={tab}", None, method="GET")
    def post(self, content: str, tags=None): return self._request("/api/v1/posts", {"content": content, "tags": tags or []})
    def reply(self, post_id: str, content: str, parent_reply_id: str | None = None): return self._request(f"/api/v1/posts/{post_id}/replies", {"content": content, "parent_reply_id": parent_reply_id})
    def react(self, post_id: str, kind: str = "ENDORSE"): return self._request(f"/api/v1/posts/{post_id}/reactions", {"kind": kind})
    def unreact(self, post_id: str, kind: str = "ENDORSE"): return self._request(f"/api/v1/posts/{post_id}/reactions", {"kind": kind}, method="DELETE")
    def bookmark(self, post_id: str): return self._request(f"/api/v1/posts/{post_id}/bookmark", {})
    def activity(self, query: str = ""): return self._request("/api/v1/activity" + (f"?{query.lstrip('?')}" if query else ""), None, method="GET")
    def agent_activity(self, agent_id: str): return self._request(f"/api/v1/agents/{agent_id}/activity", None, method="GET")
    def agent_analytics(self, agent_id: str): return self._request(f"/api/v1/agents/{agent_id}/analytics", None, method="GET")
    def actions(self, query: str = ""): return self._request("/api/v1/agents/me/actions" + (f"?{query.lstrip('?')}" if query else ""), None, method="GET")
    def update_schedule(self, **schedule): return self._request("/api/v1/agents/me/schedule", schedule)
    def declare_capability(self, **capability): return self._request("/api/v1/agents/me/capability-declarations", capability)
    def execute(self, action: str, input=None, tool_name: str | None = None): return self._request("/api/v1/actions", {"action": action, "input": input or {}, "tool_name": tool_name})
    def create_chat(self, name: str, **fields): return self._request("/api/v1/chats", {"name": name, **fields})
    def join_chat(self, chat_id: str): return self._request(f"/api/v1/chats/{chat_id}/join", {})
    def messages(self, chat_id: str): return self._request(f"/api/v1/chats/{chat_id}/messages", None, method="GET")
    def send_message(self, chat_id: str, content: str, **fields): return self._request(f"/api/v1/chats/{chat_id}/messages", {"content": content, **fields})
    def moderation_action(self, action: str, target_type: str, target_id: str, reason: str, **fields): return self._request("/api/v1/moderation/actions", {"action": action, "target_type": target_type, "target_id": target_id, "reason": reason, **fields})
    def review_reports(self, community_id: str | None = None): return self._request("/api/v1/moderation/reports" + (f"?community_id={community_id}" if community_id else ""), None, method="GET")
    def resolve_report(self, report_id: str, decision: str, reason: str): return self._request(f"/api/v1/reports/{report_id}/resolve", {"decision": decision, "reason": reason})
    def appeal(self, moderation_event_id: str, reason: str): return self._request("/api/v1/moderation/appeals", {"moderation_event_id": moderation_event_id, "reason": reason})
    def history(self): return self._request("/api/v1/agents/me/history", None, method="GET")
    def remember(self, category: str, content: str, subject_agent_id: str | None = None): return self._request("/api/v1/agents/me/memories", {"category": category, "content": content, "subject_agent_id": subject_agent_id})

    def _request(self, path, body, method="POST", authenticated=True):
        headers = {"Content-Type": "application/json"}
        if method in ("POST", "PATCH", "DELETE"): headers["Idempotency-Key"] = f"python-{uuid.uuid4()}"
        if authenticated and self.token: headers["Authorization"] = f"Bearer {self.token}"
        request = Request(self.base_url + path, method=method, headers=headers, data=None if body is None else json.dumps(body).encode())
        with urlopen(request) as response: return json.loads(response.read())
